# Birdseed2geno.r is a long script to do a lot of related things
# 1. Read the latest Affy annotation file to filter non dbSNP
#    probeset, and generate plink map file.  
# 2. Collect clinical data to combine and write a patient - gender
#    - vitalstatus file
# 3. Modify SNP annotation file to include two base genotyp
#    representation.
# 4. Read pre-collected birdseed genotyping matrix
#    (row as probeset, and column as file name), convert column names 
#    to be aliquot barcode, filter out those without clinical data (gender)
#    and not in [disease].matched.aliquots file, and then filter out 
#    non dbSNP SNPs.  The outputs are [disease].geno.rda files and summary
#    statistics geno.rda.stats file


options(stringsAsFactors=F)
require(RCurl)
source("~/github/eqtl/module.access.r")
source("~/github/eqtl/module.annotation.r")

# read sdrf list
sdrf.list <- read.delim("./snp/sdrf.list", h=F, stringsAsFactors=F)

# read diseases
diseases = unlist(read.delim("diseases.txt", h=F))

# read Affy SNP6 annotation data which is generated by 
# ExtractAffyAnnot.r on GenomeWideSNP_6.na35.annot.csv
annot = read.delim("./meta/na35.CEU.txt", h=T)

# read TCGA SNP6 birdseed probeset in order
probeset = unlist(read.table("./meta/birdseed.probeset.txt", h=F))

# filter annotation by probeset 
m = match(probeset, annot$probeset)
annot = annot[m, ]

# filter annotation by dbSNP id
w = which(substr(annot$dbSNP, 1,2) != "rs")
annot = annot[-w,]

# Generate and output plink map file
map = with(annot, cbind(chr, dbSNP, "0", pos))
colnames(map) = c("chr", "dbSNP", "dist", "pos")
map = data.frame(map)
write.table(map, "./plink/tcga.map", sep="\t", quote =FALSE, col.names=FALSE, row.names=FALSE)

# Extract gender information from clinical data
clin.summary = NULL

for(disease in diseases) {
  # read clinical information from firehose downloaded files
  clin.file = paste0("./clin/", disease, ".clin.merged.picked.txt")
  clin = read.delim(clin.file, quote="\"", row.names=1, as.is=T)
  colnames(clin) = gsub("\\.", "-", toupper(colnames(clin)))
  clin = data.frame(t(clin))

  # Extract gender information. male:1, female:2, other:-9
  gender = tolower(clin$gender)
  gender[which(gender=="male")] = "1"
  gender[which(gender=="female")] = "2"
  gender[which(! gender %in% c(1,2))] = "-9"
  
  # make output
  patient = rownames(clin)
  output = data.frame(patient)
  output$gender = gender
  output$dead = clin$vital_status
  clin.summary = rbind(clin.summary, output)
}
# write the result to file
# write.table(clin.summary, "./aliquot/clin.summary.txt", col.names=T, row.names=F, sep="\t", quote=F)

# Prepare annotation for ped format output
annot$code0 = paste(annot$Allele.A, annot$Allele.A, sep=" ")
annot$code1 = paste(annot$Allele.A, annot$Allele.B, sep=" ")
annot$code2 = paste(annot$Allele.B, annot$Allele.B, sep=" ")

# write.table(annot, "./meta/tcga.snp.annotation.txt", col.names=T, row.names=F, sep="\t", quote=F)

# initialize summary stats
output = data.frame(matrix(nrow=nrow(sdrf.list), ncol=3))
colnames(output) = c("disease", "numSample", "numSNP")

for(i in 1:nrow(sdrf.list)) {

  # Extract disease and sdrf info
  disease = sdrf.list$V1[i]
  sdrf.link = sdrf.list$V2[i]  
  sdrf <- GetTCGATable(sdrf.link)
  file.info <- ProcessSNP6Sdrf(sdrf, disease)
  file.info <- file.info[substr(file.info$aliquot, 14, 15) == "10", ]  
  aliquot = file.info$aliquot
  
  # Get Genotype data
  birdseed.file = paste0("./snp/", disease, ".birdseed.rda")
  load(birdseed.file)

  # change file name to aliquot barcode
  colnames(geno) = aliquot
  
  # read matched aliquots and patients data, and filter genotype
  match.file = paste0("./aliquot/", disease, ".matched.aliquots")
  match = read.delim(match.file, h=T)
  geno = geno[, which(colnames(geno) %in% match$SNP)]
  
  # filter by clinical info
  clin = merge(x=match, y=clin.summary, by.x="patients", by.y="patient")
  clin = clin[which(clin$gender %in% c(1,2)), ]
  m = match(clin$SNP, colnames(geno))
  geno = geno[,m]
  
  # filter out non dbSNP probeset
  m = match(annot$probeset, rownames(geno))
  geno = geno[m, ]
  rownames(geno) = annot$dbSNP
  
  # output geno.rda 
  geno.file = paste0("./snp/", disease, ".geno.rda")
  save(geno, file=geno.file)  
  
  # Collect stats
  output[i, ] = c(disease, ncol(geno), nrow(geno))
}
  
write.table(output, "./snp/geno.rda.stats", col.names=T, row.names=F, quote=F, sep="\t")

